src/java/org/apache/cassandra/io/compress/ZSTDCompressor.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.compress;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.github.luben.zstd.Zstd;
+
+public class ZSTDCompressor implements ICompressor
+{
+    public static final int FAST_COMPRESSION = 1; // fastest compression time
+    public static final int DEFAULT_LEVEL = 3;
+    public static final int BEST_COMPRESSION = 22;// very good compression ratio
+    private static final ZSTDCompressor instance = new ZSTDCompressor();
+    private static final String COMPRESSION_LEVEL_OPTION_NAME = "compression_level";
+    @VisibleForTesting
+    protected static final int compressionLevel = DEFAULT_LEVEL;
+
+    public static ZSTDCompressor create(Map<String, String> compressionOptions)
+    {
+        validateCompressionLevel(parseCompressionLevelOption(compressionOptions));
+        return instance;
```

Comment:
If we want to support different sstables/column families having different levels I think that this needs to be something closer to what the LZ4 compressor does where it creates a `ConcurrentHashMap` of instances keyed by parameter rather than just one instance. In practice I _think_ this will lead to every compressor having the default level (might want to TDD and ensure that different parameters yield different instances).

Reply:
Yes, right now the configuration parameter isn't used at all...
Regarding the ConcurrentHasMap - it makes sense only when [explicit context](http://facebook.github.io/zstd/zstd_manual.html#Chapter5) is used, but this functionality isn't exposed by JNI bindings yet.

Reply:
Sorry, just to clarify this is a Cassandra level thing, not a Zstd level thing. Cassandra calls the `create` method via reflection [here](https://github.com/apache/cassandra/blob/caf50de31b034ed77140b3c1597e7ca6ddc44e17/src/java/org/apache/cassandra/schema/CompressionParams.java#L288-L289), and since the `ICompressors` are "stateless" we can make a reasonable cache using a `ConcurrentHashMap` like [LZ4Compressor](https://github.com/apache/cassandra/blob/06209037ea56b5a2a49615a99f1542d6ea1b2947/src/java/org/apache/cassandra/io/compress/LZ4Compressor.java#L61) does keyed by parameter. This assumes of course that the `Zstd` apis are threadsafe, but I'm pretty confident they are.

Reply:
@jolynch @alexott indeed compression_level is currently always set to default, I will refactor this a bit so that each CF will have CompressionParams with dedicated compressor by removing the static reference. 
Seems to me that there is a tradeoff between  having static reference with statefulness based on compression_level versus just creating multiple compressor with each CompressionParams.
Having individual compressor might increase the number of compressor objects floating around but we have the flexibility of directly using compression_level without any complicated pre-determined conditions. let me know what you think.

Reply:
From my brief reading of the JNI Bindings it seems that zstd-jni uses `ZSTD_compress` and not `ZSTD_compressCCtx`. The latter allows you to pass in a context which would've been more efficient. For now, I think it would be ok to simply pass in the compression level. The only value of keeping multiple compressor objects around is to retain the `CompressionParams`. It would be useful to have a caching factory to avoid creating multiple objects with the same compression params.

As far as thread safety goes, the JNI Code looks thread safe specifically compress and decompress methods. As pointed out earlier they create and destroy the Compression Context on each invocation which is memory unfriendly. So, although the zstd benchmarks may look great, I am not so sure about this JNI binding.

We should definitely add a JMH Benchmark for zstd.